sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs

```diff
@@ -96,48 +96,52 @@ private ExpressionSyntax GetCallbackArg(InvocationExpressionSyntax invocationExp
             return callbackArg;
         }
 
-        private static bool IsCallbackMayContainEndInvoke(SyntaxNode callbackArg, SemanticModel semantic)
+        /// <returns>
+        /// false if callback code has been resolved and does not contain "EndInvoke",
+        /// true if callback code contains "EndInvoke" or callback code has not been resolved.
+        /// </returns>
+        private static bool CallbackMayContainEndInvoke(SyntaxNode callbackArg, SemanticModel semantic)
         {
-            var calledMethod = callbackArg.RemoveParentheses();
-            calledMethod = ReplaceIdentifierByItsInitializer(calledMethod, semantic);
-            calledMethod = ReplaceNewDelegateByMethodDeclaration(calledMethod, semantic);
-            if (calledMethod != null && ParentDeclarationKinds.Contains(calledMethod.Kind()))
+            var callback = callbackArg.RemoveParentheses();
+            if (callback.IsKind(SyntaxKind.IdentifierName))
             {
-                return GetEndInvokeList(calledMethod, semantic).Count > 0;
+                callback = LookupIdentifierInitializer((IdentifierNameSyntax)callback, semantic);
             }
 
-            return true;
-        }
-
-        private static SyntaxNode ReplaceNewDelegateByMethodDeclaration(SyntaxNode node, SemanticModel semantic)
-        {
-            if (node != null && node.IsKind(SyntaxKind.ObjectCreationExpression))
+            if (callback.IsKind(SyntaxKind.ObjectCreationExpression))
```

Comment:
should this be an `else if`? or are you checking the result of `LookupIdentifierInitializer`?